presto-raptorx/src/main/java/com/facebook/presto/raptorx/util/PrioritizedFifoExecutor.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptorx.util;
+
+import com.google.common.collect.ComparisonChain;
+import com.google.common.util.concurrent.ExecutionList;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.log.Logger;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is based on {@link io.airlift.concurrent.BoundedExecutor}.
+ */
+@ThreadSafe
+public class PrioritizedFifoExecutor<T extends Runnable>
```

Comment:
Good idea, let me try that

Reply:
I tried that, but there are a couple reasons against it:

1. `PrioritizedFifoExecutor` is supposed to be generic, so it needs to accept a `Comparator`. It could accept both, but it's easy enough for callers to do `Comparator.naturalOrder()`.
2. Using a `Comparator` actually seems cleaner, since it can be implemented using `comparingInt()`.